src/python/pants/fs/fs_test.py

```diff
@@ -0,0 +1,39 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from pathlib import Path
+
+from pants.engine.fs import (Digest, DirectoryToMaterialize, FileContent, InputFilesContent,
+                             MaterializeDirectoriesResult, MaterializeDirectoryResult, Workspace)
+from pants.util.contextutil import temporary_dir
+from pants_test.test_base import TestBase
+
+
+class FileSystemTest(TestBase):
```

Comment:
Writing a tiny `@console_rule` inline at the top level of this file might be a cleaner test... it would also have the benefit of getting you an anonymous build root at `self.build_?root` to avoid the need for the tmpdir. But not a blocker.

Reply:
@pierrechevalier83 and I tried to actually use this in an `@console_rule` and we found that it panic'd because it was calling `block_on` on the tokio Executor from a Future that was already being `block_on`'d (the Future driving the whole `@console_rule` execution).

Could you add an actual `@console_rule` that uses this, to show it works in an integration test (and fix what needs fixing to make that pass)?

Reply:
Ah. Ok: a blocker then.

Reply:
Winding back a bit to the 1st cut at this that exposed an Intrinsic and in spirit with solving the current re-entrancy tangle:

Given:
1. Intrinsic rules learn the cacheable bit that Task rules know: https://github.com/gshuflin/pants/blob/13687070b3bcd08eb23eec5d256b8f90da49fd5c/src/rust/engine/src/tasks.rs#L238
2. A rule graph validation that checks a cacheable nose has no uncacheable node inputs.

Would this not safely solve the issues here? Fwict it would enforce console rules could be the only users of the intrinsic in particular, but more generally would (safely) allow arbitrarily large paths from the root of the rule graph to be uncacheable as long as the boundary defining the transition from the uncacheable portion of the graph to the cacheable portion is the only such one.

Reply:
> Would this not safely solve the issues here? Fwict it would enforce console rules could be the only users of the intrinsic in particular, but more generally would (safely) allow arbitrarily large paths from the root of the rule graph to be uncacheable as long as the boundary defining the transition from the uncacheable portion of the graph to the cacheable portion is the only such one.

It would solve the "how to execute" question, but IMO it would confuse semantics: So far everything that we `yield` for is referentially transparent and sideeffect free... this is discussed on #8347 as well. Opening the door to "yielding might have sideeffects" is a big, wide door. So as with #8347, it's something to do very cautiously.

I think that there is a fairly straightforward fix for this issue that was discussed in Slack, so if we can avoid opening that door (yet, perhaps?) that would be good.